orchestra/model/component.py

```diff
@@ -53,13 +62,52 @@ def branch(self):
 
     @property
     def recursive_hash(self):
-        if self._recursive_hash is None:
-            raise Exception("Accessed recursive_hash before calling resolve_dependencies")
+        assert self._recursive_hash is not None, "Accessed recursive_hash before calling resolve_dependencies"
         return self._recursive_hash
 
+    @cache
+    def recursive_hash_material(self) -> str:
+        """Returns the string that is hashed to compute recursive_hash"""
+        hash_material = None
+
+        if self._use_config_cache:
+            hash_material = self._get_cached_hash_material()
+
+        if hash_material is None:
+            components_to_hash = list(self._transitive_dependencies())
+            components_to_hash.sort(key=lambda c: c.name)
+            hash_material = [c.serialize() for c in components_to_hash]
+
+            hash_material = yamldump(hash_material)
+            self._cache_hash_material(hash_material)
+
+        return hash_material
+
+    def _get_cached_hash_material(self) -> Optional[str]:
+        if self._cache_filepath.exists():
+            with open(self._cache_filepath) as f:
+                cached_config_hash, _, cached_commit = next(f).strip().partition(" ")
+                wanted_commit = self.commit() or ""
+                if cached_config_hash == self._configuration_hash and cached_commit == wanted_commit:
+                    return f.read()
+        return None
+
+    def _cache_hash_material(self, hash_material: str):
+        os.makedirs(self._cache_filepath.parent, exist_ok=True)
+        with open(self._cache_filepath, "w") as f:
+            f.write(f"{self._configuration_hash} {self.commit() or ''}\n")
+            f.write(hash_material)
+
+    @property
+    def _cache_filepath(self) -> Path:
+        return Path(self._orchestra_cache_dir) / "hash_material" / self._name_for_hash_material_cache
```

Comment:
So we have a file for each component? Works.
At this point, how hard is it to make it so that we can `md5sum` (or whatever hash you're using) and get as a result the name of the binary archive?

```suggestion
        return Path(self._orchestra_cache_dir) / "hash-material" / self._name_for_hash_material_cache
```

Reply:
Yes, one file per component is easier and faster to work with than a single gigantic file.
Using `sha1sum` on the hash material file stored alongside the binary archive gives the recursive hash (part of the filename, which also includes the commit).